data/transactions/logic/opcodes.go

```diff
@@ -507,6 +509,34 @@ func opMatchStackChange(cx *EvalContext) (deletions, additions int) {
 	return
 }
 
+// CurrentScratchChange tells if current opcode is storing value to a scratch slot,
+// and if so, it returns scratch slot id, previous value, new value to be written;
+// otherwise, it indicates the current opcode is not a scratch slot chagne.
+func (cx *EvalContext) CurrentScratchChange() (scratchSlot uint64, newValue basics.TealValue, isScratchChange bool) {
+	currentOpcodeName := opsByOpcode[cx.version][cx.program[cx.pc]].Name
+	last := len(cx.Stack) - 1
+
+	switch currentOpcodeName {
+	case "store":
+		scratchSlot = uint64(cx.program[cx.pc+1])
+	case "stores":
+		prev := last - 1
+		scratchSlot = cx.Stack[prev].Uint
+
+		// If something goes wrong for `stores`, we don't have to error here
+		// for in runtime already has evalError
+		if scratchSlot >= uint64(len(cx.scratch)) {
+			return
+		}
+	default:
+		return
+	}
+
+	newValue = cx.Stack[last].ToTealValue()
+	isScratchChange = true
+	return
```

Comment:
I think it's clearer to avoid the naked return here:
```suggestion
	return scratchSlot, cx.Stack[last].ToTealValue(), true
```